### ISSUE TITLE
List.memberOf returns incorrect results for NaN

### ISSUE THREAD

**Issue by soc** [flix/flix#3843]
def main(): Unit & Impure =
     // prints `false`, should print `true`
    List.memberOf(0.0/0.0, (0.0/0.0) :: Nil) |> println

The root cause is that, despite `Eq` being declared a lawful typeclass with reflexivity/transitivity laws, the `Eq` instance for floats does not use the correct comparison operation.

It uses an operation similar or equal to the one described in IEEE754 §5.11, but for reflexivity/transitivity to hold, it needs to implement `$FLOAT64_EQ$`/`$FLOAT64_NEQ$` similar or equal to IEEE754 §5.10.

**Comment by magnus-madsen** [comment]
Agreed. This revives our discussion of whether to have PartialEq and Eq.

**Comment by soc** [comment]
I'd guess it's largely a solved problem these days.

I think going the `Eq`/`Id` / `Equals`/`Identity` route is helpful to get the mental model right for readers (i. e. that there is no subtyping relationship between those two).

**Comment by magnus-madsen** [comment]
Could you expand on that or do you have a reference?

**Comment by soc** [comment]
Sure!

The core insight is that `PartialEq`/`Eq` usually implies some form of sub-typing relationship with the expectation that the partial versions are "law-reduced", but *have compatible semantics*. For floating points, this is not the case and it's intentionally designed this way. 

So what I usually do is to cleanly separate a notion of equality that is domain-specific (defined by the type itself), and a notion of identity (that is solely defined by a value's bits).

Operations then need to decide which typeclass (or both) they are actually interested in to provide the semantics expected by users.

I documented this approach a while ago [Part 1](https://soc.me/languages/equality-and-identity-part1), [Part 2](https://soc.me/languages/equality-and-identity-part2), [Part3](https://soc.me/languages/equality-and-identity-part3) and [Part4](https://soc.me/languages/equality-and-identity-part3).

Here is [another language](https://github.com/dinfuehr/dora/blob/main/dora/stdlib/traits.dora) that successfully uses this design.

**Comment by magnus-madsen** [comment]
Interesting. We should take a closer look.

Quick question: Which equality is then used for e.g. a Set?

**Comment by soc** [comment]
> Which equality is then used for e.g. a Set?

Depends on the intended use-case.

In the past I provided a `contains[T : Identity & Equality]` and a more primitive `has[T : Identity]` function.

**Comment by magnus-madsen** [comment]
@mlutze @JonathanStarup We should discuss this at some point.

**Comment by soc** [comment]
In terms of code (excuse my Flix-like pseudo-code), the design looks somewhat like that:

```
pub lawful class Identity[a]
  fun id(this: a, that: a) ...
  fun nid(this: a, that: a) ...
  // laws: reflexivity, symmetry, transitivity

pub lawful class Equality[a] ...
  fun eq(this: a, that: a) ...
  fun neq(this: a, that: a) ...
  // laws: symmetry

pub lawful class Sorting[a] extends Identity[a]
  fun sortsAbove ...
  fun sortsBelow ...

pub lawful class Comparing[a] extends Equality[a]
  fun lessThan ...
  fun greaterThan ...
```

with instances as expected:

```
instance Identity[Float64] {
    pub def id(x: Float64, y: Float64): Bool = $FLOAT64_ID$(x, y) // IEE754 §5.10
    pub override def nid(x: Float64, y: Float64): Bool = $FLOAT64_NID$(x, y) // IEE754 §5.10
}

instance Equality[Float64] {
    pub def eq(x: Float64, y: Float64): Bool = $FLOAT64_EQ$(x, y) // IEE754 §5.11
    pub override def neq(x: Float64, y: Float64): Bool = $FLOAT64_NEQ$(x, y) // IEE754 §5.11
}
```
